nntrainer/tensor/cpu_backend/x86/avx2_impl.cpp

```diff
@@ -12,12 +12,298 @@
+#define UNLIKELY [[unlikely]]
+#else
+#define UNLIKELY
+#endif
+
+#if !defined(_MSC_VER) && !defined(__clang__)
+#pragma GCC diagnostic ignored "-Wattributes"
+#endif
+
+namespace {
+
+template <typename To_, typename From_>
+constexpr inline bool concept17_BinaryCastable =
+  sizeof(To_) == sizeof(From_) &&
+  std::is_trivially_copyable_v<From_> &&std::is_trivially_copyable_v<To_>;
+
+template <class To_, class From_>
+auto compat_bit_cast(const From_ &src) noexcept
+  -> std::enable_if_t<concept17_BinaryCastable<To_, From_>, To_> {
+#if __cpp_lib_bit_cast >= 201806L
+  return std::bit_cast<To_>(src);
+#else
+  To_ dst;
+  std::memcpy(&dst, &src, sizeof(To_));
+  return dst;
+#endif
+}
+
+[[nodiscard]] constexpr inline unsigned
+constexpr_popcount(uint32_t v) noexcept {
+#if __cpp_lib_bitops >= 201907L
+  return std::popcount(v);
+#else
+  // Popcount via bit-hack
+  v = v - ((v >> 1) & 0x55555555);                // reuse input as temporary
+  v = (v & 0x33333333) + ((v >> 2) & 0x33333333); // temp
+  auto c = (((v + (v >> 4)) & 0xF0F0F0F) * 0x1010101) >> 24; // count
+  return c;
+#endif
+}
+
+template <unsigned I_>
+constexpr inline bool concept17_PowerOfTwo = (constexpr_popcount(I_) == 1);
+
+namespace numbers {
+
+#if __has_include(<numbers>) && __cpp_lib_math_constants >= 201907L
+using std::numbers::ln2_v;
+using std::numbers::log2e_v;
+#else
+template <typename Float_> constexpr inline auto ln2_v = Float_{M_LN2};
+
+template <typename Float_> constexpr inline auto log2e_v = Float_{M_LOG2E};
+#endif
+} // namespace numbers
+
+constexpr inline float EXP_ARG_MIN = -87.0;
+constexpr inline float EXP_ARG_MAX = +88.3762626647949f;
+
+// @brief Precalculated lookup table for 2^x calculation
+template <unsigned N_, typename Ty_ = uint32_t, typename Float_ = float,
+          typename = std::enable_if_t<concept17_PowerOfTwo<N_>>>
+struct Exp2Table {
+
+  constexpr static inline auto MANTISSA_BITS =
+    std::numeric_limits<Float_>::digits - 1;
+
+#if __cpp_consteval >= 201811L && __has_constexpr_builtin(__builtin_exp2)
+  [[nodiscard]] static consteval auto calculate() noexcept {
+    std::array<Ty_, N_> t;
+    for (unsigned i = 0; i < N_; ++i)
+      t[i] = std::bit_cast<Ty_>(std::exp2(Float_{1.0} * i / N_)) -
+             ((i << MANTISSA_BITS) / N_);
+    return t;
+  }
+#endif
+};
+
+#if !__has_constexpr_builtin(__builtin_exp2) || !(__cpp_consteval >= 201811L)
+
+// @brief Precalculated lookup table for 2^x calculation when we don't have
+// constexpr math functions
+template <> struct Exp2Table<8, uint32_t, float> {
+  [[nodiscard]] static constexpr auto calculate() noexcept {
+    std::array<uint32_t, 8> t = {0x3f800000U, 0x3f7b95c2U, 0x3f7837f0U,
+                                 0x3f75fed7U, 0x3f7504f3U, 0x3f75672aU,
+                                 0x3f7744fdU, 0x3f7ac0c7U};
+    return t;
+  }
+};
+
+#endif
+
+template <unsigned N_> // requires PowerOfTwo<N_>
+inline constexpr auto exp2_table_v = Exp2Table<N_>::calculate();
```

Comment:
requires aligned_array or `alignas(16)`

Reply:
https://gist.github.com/piotrrak/1e5c5026358c21d0038495535295c740 aligned_array implementation, but it is c+20, so alignas for now.